Always detach volumes in operator executor

jsravn opened the pull request:
**What this PR does / why we need it**:

Instead of marking a volume as detached immediately in Kubelet's
reconciler, delegate the marking asynchronously to the operator
executor. This is necessary to prevent race conditions with other
operations mutating the same volume state.

An example of one such problem:

1. pod is created, volume is added to desired state of the world
2. reconciler process starts
3. reconciler starts MountVolume, which is kicked off asynchronously via
   operation_executor.go
4. MountVolume mounts the volume, but hasn't yet marked it as mounted
5. pod is deleted, volume is removed from desired state of the world
6. reconciler reaches detach volume section, detects volume is no longer in desired state of world,
   removes it from volumes in use
7. MountVolume tries to mark mount, throws an error because
   volume is no longer in actual state of world list. After this, kubelet isn't aware of the mount
   so doesn't try to unmount again.
8. controller-manager tries to detach the volume, this fails because it
   is still mounted to the OS.
9. EBS gets stuck indefinitely in busy state trying to detach.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #32881, fixes ##37854 (maybe)

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40417)
<!-- Reviewable:end -->

jsravn commented:
@jingxu97 can you have a look?

jingxu97 commented:
@jsravn thanks a lot for the PR. But I think pause desired state updating may not (completely) solve the issue. From the log you posted in #32881, I can see the race condition is as follows. Please note that in default setup, volume reconciler running in master is responsible for attaching volumes, and the reconciler in kubelet node is responsible for mounting global device path and bind mount.
1. Add pod with attachable volume, reconciler on master is trying to attach the volume to the node, reconciler on kubelet is verifying whether volume is attached or not.
2. Once the volume is attached to the node, reconciler on kubelet will put this volume into actual state's attached volume list. Reconciler issue asynchronous call to mount global device.
3. At this moment, this pod is deleted and removed from desired state. So reconciler checks since it is not yet mounted so it is ok to delete the volume from the attached volume list in actual state. 
4. The asynchronous call for mounting global device finishes successfully (mount to device path). But when it tries to update the actual state about this mount information, it fails because there is no such volume in attached list in actual state any more (caused by step 3). 
At this state, reconciler actual state shows that the volume is not attached or mounted (but actually it is mounted already), so reconciler at kubelet will not issue any unmount operation. When master node reconciler decides to detach after timeout, the volume might stuck in busy state.

Your solution to pause updating desired state might help (step 3 pod will not be removed from desired state in the same loop), but since step 2 is asynchronous call which might take arbitrary time to finish. So desired state might get updated in the next reconciler loop and in step 4 the same issue might happen. Also in theory, because of this pause, desired state might never get a chance to update. Both reconciler and desired state population runs periodically.

jsravn commented:
@jingxu97 You're right, the race is still there, but it would only happen if the mount operation took longer than the reconciler sync period, when in practice it happens very fast (just attach is slow). There are a lot of approaches here to solve the problem - all state changes could go through operation_executor, or we could make any calls in reconciler synchronous (such as waiting for the mount to complete). I suspect there are other race conditions in here as well, in particular if the volume-controller-manager is disabled I suspect it can get in a bad state if a pod is deleted during the reconcile process (but I haven't spent enough time looking at the code to verify this). I thought simply pausing updates to desired state of the world would cover most of these edge cases. Perhaps combining that w/ synchronous mounts/unmounts would solve all the problems.

jingxu97 commented:
@jsravn, the reconciler sync period is 100ms, which is also very short. We don't want to use synchronous mount/umounts due to performance issue. There might be a large number of mounts in some setup. 
Now I am thinking to add one checking for pending operation should good enough to solve the issue. 
if rc.operationExecutor.IsOperationPending() {
   continue
}
After adding this check, in the above step 3, since there is a mounting operation pending, the reconciler will not delete the volume from the attached volume list in actual state. Only if there is no pending operation, it will check whether the volume is already mounted or not and determine whether need to unmount first.

jsravn commented:
> @jsravn, the reconciler sync period is 100ms

Ahh right, I was thinking of the volume manager's period.

jsravn commented:
@jingxu97 Considering it has a 100ms sync period I agree with you that my solution isn't going to work. I was looking into what you suggested. One problem at least is we don't have the pod name (it has been deleted at this point), so we'd have to add some additional code that checks for pending operations against a volume only. I think that would fix the problem I described. But the code is getting quite complex so it's hard to tell for sure.

jsravn commented:
@jingxu97 are you going to work on a fix? Otherwise I can keep this open and make the changes myself.

jsravn commented:
I think the correct approach is to add a new method in operation_executor called `VerifyControllerDetachedVolume`, which mirrors `VerifyControllerAttachedVolume`. This should ensure the marking detached happens in an orderly fashion. I'm working on it now.

jsravn committed:
Mark detached only if no pending operations

To safely mark a volume detached when the volume controller manager is used.

An example of one such problem:

1. pod is created, volume is added to desired state of the world
2. reconciler process starts
3. reconciler starts MountVolume, which is kicked off asynchronously via
   operation_executor.go
4. MountVolume mounts the volume, but hasn't yet marked it as mounted
5. pod is deleted, volume is removed from desired state of the world
6. reconciler detects volume is no longer in desired state of world,
   removes it from volumes in use
7. MountVolume tries to mark volume in use, throws an error because
   volume is no longer in actual state of world list.
8. controller-manager tries to detach the volume, this fails because it
   is still mounted to the OS.
9. EBS gets stuck indefinitely in busy state trying to detach.

jsravn commented:
@jingxu97 I've made the changes, have a look. Thanks :)

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -820,6 +823,54 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
+		}
+
+		// Fetch current node object
+		node, fetchErr := og.kubeClient.Core().Nodes().Get(string(volumeToDetach.NodeName), metav1.GetOptions{})
```
I think this is a good solution by putting this function in to operation_executor. But I feel this Fetch current node object might cause unnecessary overhead by querying API server frequently. It is ok to remove this fetch node object part and just call MarkVolumeAsDetached() function. The reason is that before calling this  VerifyControllerDetachedVolume(), reconcilder already checked that attachedVolume.GloballyMounted is false. This means that there is no global mount (mount a device path to a global path) exists or there is a pending operation for this volume being mounted. However, if there is an pending operation for this volume, operation_executor will not start this   VerifyControllerDetachedVolume() at all. It must wait until the pending operation for this volume finishes, it can start this verify detached volume operation.

jsravn commented:
@jingxu97 Okay I made the change. I left in the log message though so we can know when Kubelet considers the volume detached for debugging future issues.

jsravn commented:
@saad-ali is it possible to run the build on this? Would be nice to get this fix soonish, as it breaks EBS for us quite a lot.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -820,6 +823,27 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
+			volumeToDetach.VolumeSpec.Name(),
+			volumeToDetach.DevicePath)
+		actualStateOfWorld.MarkVolumeAsDetached(volumeToDetach.VolumeName, volumeToDetach.NodeName)
+		return nil
```
I think we have to put the logic of checking whether volume is globally mounted or not inside of this function. Otherwise there might be a race condition. When reconciler checks globalMounted is false, and then right after that, this state is changed to true. When calling VerifyControllerDetachVolume it will think it is not mounted and mark it as detached.

jsravn replied:
Ok, I'll have a look.

jsravn replied:
Yeah you're right. My previous idea of checking the node status for list of attached volumes would have prevented this, though. I will move the globally mounted check inside as well.

jsravn replied:
It's difficult to get the globally mounted state inside of operation_executor, since it's  private state of the kubelet's actualStateOfTheWorld. We need to requery the asw inside of the callback to check if globally mounted changed (can't just pass it in from reconciler, since it's a copy). I can either open up the volume's asw interface, or can check if the device is mounted directly. I'm inclined to do the latter, since it's less dependent on state management in kubelet, and probably a more reliable check of the real state of things.

jsravn replied:
Note that `VerifyControllerAttachedVolumeFunc` queries the apiserver to see if the volume is attached. How is that different from my original solution of checking the node status to see if the volume is detached? Should we fix that as well?

jingxu97 replied:
Yes, VerifyControllerAttachedVolumeFunc check node status. But we could not change this because device path for some volume plugins such as aws needs to be passed by node status.

jsravn commented:
@jingxu97 Have a look at my latest change. I'm using the mounter now to check on the local system if the volume is detached, by checking if the device is still present or not. This doesn't incur any external calls and should reliability allow us to verify the device has been globally unmounted and detached. I was thinking we could use `mounter.DeviceOpened` instead, but this is better I think since it waits for the device to really be detached.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -820,6 +823,44 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
+			return nil
+		}
+
+		attached, err := mounter.PathIsDevice(volumeToDetach.DevicePath)
```
Just a note that PathIsDevice check might hang for nfs mount when server is down. Since this check is in a gorountine, it should not affect other operations. NFS hanging problem is still under investigation

jingxu97 commented:
/approve

thockin commented:
/approve

saad-ali reviewed (changes requested):
A much cleaner approach would be to wrap `rc.actualStateOfWorld.MarkVolumeAsDetached` in a `operationExecutor.IsOperationPending` check. That way you can avoid adding a new operation, that's not really an operation, to `operation_executor`.

saad-ali commented:
If you can get this change merged by Tuesday and cherry picked to 1.4 and 1.5 branches by Wed, we can get this it into the 1.5.3 and 1.4.9 releases.

saad-ali commented:
/approve cancel

jsravn commented:
> A much cleaner approach would be to wrap rc.actualStateOfWorld.MarkVolumeAsDetached in a operationExecutor.IsOperationPending check. That way you can avoid adding a new operation, that's not really an operation, to operation_executor.

I'm not sure it's cleaner, since at that point you don't have the pod name anymore which IsOperationPending requires. So now the reconciler will have to pass in an empty pod string, with the implicit understanding that that becomes a wildcard match in the operation_executor (need to check this works, just thinking from memory). Plus, all other asw state mutation happens in operation_executor iirc, so I think doing MarkVolumeAsDetached in the reconciler is bad idea (which is why this bug happened in the first place). I don't like it but I will try it out tomorrow and try to make sure there are no problems with the approach.

saad-ali commented:
> I'm not sure it's cleaner, since at that point you don't have the pod name anymore which IsOperationPending requires. So now the reconciler will have to pass in an empty pod string, with the implicit understanding that that becomes a wildcard match in the operation_executor (need to check this works, just thinking from memory). 

Yes, you would pass in an empty string for pod, meaning it would act as a wildcard. This would do exactly what we want: verify that there are are no pending operations for that volume before deleting it from the "Actual State" cache.

> Plus, all other asw state mutation happens in operation_executor iirc, so I think doing MarkVolumeAsDetached in the reconciler is bad idea (which is why this bug happened in the first place). I don't like it but I will try it out tomorrow and try to make sure there are no problems with the approach.

`operation_executor` doesn't do anything magical, it just runs operations asynchronously. The `nested_pending_operations` object acts as a manager for these operations. You can push a new operation in to `nested_pending_operation` or you can just query it to see if there is a pending operation. And new operations should only be created where they make sense. In general, operation executor operations should be doing something like making a network call or reading from disk. It makes sense for distinct operations like Attach, Detach, and even VerifyControllerAttachedVolume where the disk (or other signals) need to be examined to be separate operation executor operations.

But in this case all you want to do is mutate internal state based on some other internal state. So it does not make sense to have a whole separate operation. This becomes obvious if you look at the naming of your new `VerifyControllerDetachedVolume` which is incorrectly named, and moreover the check doesn't do what you want for this fix, you want to verify that there are no pending operations, instead you are checking if the path for a volume to detach is device path or not--this is a very roundabout (and possibly incorrect) way of determining this (and as Jing mentioned, could result in hangs for NFS).

jsravn commented:
@saad-ali Ok, I've made the change. It's a lot simpler.

jingxu97 commented on pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:
```diff
@@ -36,9 +36,6 @@ import (
 const (
-	// emptyUniquePodName is a UniquePodName for empty string.
-	emptyUniquePodName types.UniquePodName = types.UniquePodName("")
-
```
why delete this? Also could we make it similar to emptyUniqueVolumeName?

jsravn replied:
I moved this into `types` package since there were a few places all defining UniquePodName(""). So it's still using it, just got rid of the duplication.

jsravn replied:
@jingxu97 Not sure I understand what you mean by making it similar to emptyUniqueVolumeName - can you clarify?

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -342,10 +342,15 @@ func (rc *reconciler) reconcile() {
+							attachedVolume.VolumeName,
+							attachedVolume.VolumeSpec.Name(),
+							attachedVolume.DevicePath)
+					}
```
I think IsOperationPending() needs to be checked before attachedVolume.GloballyMounted. Otherwise there might be a race condition as follows.
1. Checked attachedVolume.GloballyMounted is false, but there is a pending operation trying to mount
2. Pending operation finishes, and volume is mounted
3. Check whether there is a pending operation, return false, then mark volume as detached.

jsravn replied:
K I will change that.

jsravn commented:
> operation_executor doesn't do anything magical, it just runs operations asynchronously. 

I misunderstood it then - I thought the purpose of it was to serialise all state updates to a single volume, since there is logic to not run updates concurrently on the same volume. I think this is a good approach - it means only a single thing is mutating the volume state, and kubelet's view of the volume state, in a relatively ordered way. It's definitely simpler to mutate the asw directly in reconcile though, just makes the logic more tricky - since you have to be aware of all the async volume changes that could be going on in the operator executor.

jsravn commented:
Ok @saad-ali and @jingxu97 can you have a look? I think I've addressed all comments.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -342,10 +344,12 @@ func (rc *reconciler) reconcile() {
-					// so just remove it to actualStateOfWorld without attach.
-					rc.actualStateOfWorld.MarkVolumeAsDetached(
-						attachedVolume.VolumeName, rc.nodeName)
+					// Kubelet not responsible for detaching or this volume has a non-attachable volume plugin.
```
Maybe put this comment above?

jsravn replied:
All the comments seem to be after the conditional so I left that as is. Do you mean move it to line 345?

jingxu97 replied:
Yes.

jsravn replied:
Done

jingxu97 commented on pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:
```diff
@@ -37,7 +37,7 @@ import (
+	EmptyUniquePodName types.UniquePodName = types.UniquePodName("")
 
 	// emptyUniqueVolumeName is a UniqueVolumeName for empty string
 	emptyUniqueVolumeName v1.UniqueVolumeName = v1.UniqueVolumeName("")
```
Can you make this Capital case too to be consistent with the podName?

jingxu97 replied:
Actually I am not sure why UniquePodName is types.UniquePodName and UniqueVolumeName is v1.UniqueVolumeName @saad-ali ?

jsravn commented:
@jingxu97 ok have another look :)

jingxu97 commented:
lgtm. @saad-ali

jsravn commented:
@saad-ali can you take a second look?

jingxu97 commented:
@saad-ali Does this PR need more approve to run the tests?

jingxu97 commented:
@k8s-bot ok to test

jsravn commented:
I've rebased/squashed the changes.

jsravn commented:
ping @saad-ali, can you have a look?

saad-ali commented:
> ping @saad-ali, can you have a look?

Taking a look

saad-ali commented:
/lgtm
/approve

Thanks @jsravn

saad-ali commented:
Please cherry pick to 1.5 branch.

gnufied commented:
@jsravn do you want to open PR for cherry picking this to `release-1.5`?

jsravn commented:
@gnufied already did https://github.com/kubernetes/kubernetes/pull/41555
